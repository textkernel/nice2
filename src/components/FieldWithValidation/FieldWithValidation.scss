.FieldWithValidation {
    margin: var(--spacing-2x) 0 0 0;
}
