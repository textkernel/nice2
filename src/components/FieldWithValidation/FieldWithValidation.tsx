import * as React from 'react';
import { Tooltip } from '../Tooltip';
import { Text } from '../Text';

interface Props {
    /** a single form field, e.g. Input, TextArea, etc. It should support context=”bad” prop */
    children: React.ReactElement;
    /** if this is not empty, this message will be rendered and context bad will be applied to the child */
    errorMessage?: string;
    /** if set to true the error message will be rendered as a tooltip. Otherwise it will be a text under the field */
    useTooltip?: boolean;
}

export const FieldWithValidation: React.FC<Props> = props => {
    const { children, errorMessage, useTooltip } = props;

    if (!errorMessage) {
        return children;
    }

    const clonedChild = React.cloneElement(children, { context: 'bad' });

    return useTooltip ? (
        <Tooltip content={errorMessage}>{clonedChild}</Tooltip>
    ) : (
        <>
            {clonedChild}
            <Text context="bad" size="small">
                {errorMessage}
            </Text>
        </>
    );
};

FieldWithValidation.defaultProps = {
    useTooltip: false,
};

FieldWithValidation.displayName = 'FieldWithValidation';
