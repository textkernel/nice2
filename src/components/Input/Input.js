import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './Input.scss';
import { CONTEXTS, INPUT_TYPES, SIZES } from '../../constants';

const { block } = bem({
    name: 'Input',
    classnames: styles,
    propsToMods: ['context', 'isBlock', 'size']
});

const Input = forwardRef((props, ref) => {
    const { children, context, disabled, isBlock, size, type, value, ...rest } = props;
    return (
        <input
            {...rest}
            {...block(props)}
            ref={ref}
            type={type}
            disabled={disabled}
            value={value}
        />
    );
});

Input.displayName = 'Input';

// Any other attributes (onChange, onKeyUp etc.) are
// supported although not defined in propTypes
Input.propTypes = {
    /** The input field context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context: PropTypes.oneOf(CONTEXTS),
    /** Should the input field be disabled or not */
    disabled: PropTypes.bool,
    /** Whether or not to show block-level input field (full width) */
    isBlock: PropTypes.bool,
    /** The size of the input field */
    size: PropTypes.oneOf(SIZES),
    /** Type of the input field */
    type: PropTypes.oneOf(INPUT_TYPES),
    /** Current input field value */
    value: PropTypes.string
};

Input.defaultProps = {
    context: 'brand',
    disabled: false,
    isBlock: false,
    size: 'normal',
    type: 'text',
    value: undefined
};

export default Input;
