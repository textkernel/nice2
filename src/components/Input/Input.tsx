import * as React from 'react';
import bem from '../../utils/bem';
import styles from './Input.scss';
import { Context, InputType, Size } from '../../constants';

// Any other attributes (onChange, onKeyUp etc.) are
// supported although not defined in props type definition
interface Props extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
    /** The input field context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context: Context;
    /** Should the input field be disabled or not */
    disabled: boolean;
    /** Whether or not to show block-level input field (full width) */
    isBlock: boolean;
    /** The size of the input field */
    size: Size;
    /** Type of the input field */
    type: InputType;
}

const { block } = bem('Input', styles);

const Input: React.FC<Props> = React.forwardRef((props, ref?: React.Ref<HTMLInputElement>) => {
    const { children, context, disabled, isBlock, size, type, value, ...rest } = props;
    return (
        <input
            {...rest}
            {...block(props)}
            ref={ref}
            type={type}
            disabled={disabled}
            value={value}
        />
    );
});

Input.displayName = 'Input';

Input.defaultProps = {
    context: 'brand',
    disabled: false,
    isBlock: false,
    type: 'text',
    size: 'normal',
};

export default Input;
