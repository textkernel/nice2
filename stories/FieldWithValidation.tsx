import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, withKnobs } from '@storybook/addon-knobs';
import { FieldWithValidation, Input, TextArea } from '@textkernel/oneui';

storiesOf('Molecules|FieldWithValidation', module)
    .addDecorator(withKnobs)
    .add('FieldWithValidation', () => (
        <>
            <div style={{ marginBottom: '10px' }}>
                <FieldWithValidation
                    errorMessage={text('Error message for input field', 'Something wrong')}
                    useTooltip={boolean('Use tooltip for input field', false)}
                >
                    <Input />
                </FieldWithValidation>
            </div>
            <div>
                <FieldWithValidation
                    errorMessage={text('Error message for text field', '')}
                    useTooltip={boolean('Use tooltip for text field', false)}
                >
                    <TextArea />
                </FieldWithValidation>
            </div>
        </>
    ));
