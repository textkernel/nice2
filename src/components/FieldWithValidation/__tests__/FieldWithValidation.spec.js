import * as React from 'react';
import toJson from 'enzyme-to-json';
import { FieldWithValidation } from '../FieldWithValidation';
import { Input } from '../../Input';

describe('FieldWithValidation', () => {
    let wrapper;

    describe('when rendering message as text', () => {
        describe('with no error message passed', () => {
            beforeEach(() => {
                wrapper = mount(
                    <FieldWithValidation>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should simply render the child as it is', () => {
                expect(wrapper.find('Input')).toHaveLength(1);
                expect(wrapper.find('Tooltip')).toHaveLength(0);
                expect(wrapper.find('Text')).toHaveLength(0);
                expect(toJson(wrapper)).toMatchSnapshot();
            });
        });
        describe('with error message passed', () => {
            const message = 'invalid field';
            beforeEach(() => {
                wrapper = mount(
                    <FieldWithValidation errorMessage={message}>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should set context to bad on child', () => {
                expect(wrapper.find('Input').prop('context')).toBe('bad');
            });
            it('should render the message as text', () => {
                expect(wrapper.find('Tooltip')).toHaveLength(0);
                expect(wrapper.find('Text')).toHaveLength(1);
                expect(wrapper.find('Text').text()).toBe(message);
            });
        });
    });
    describe('when using tooltip', () => {
        describe('with no error message passed', () => {
            beforeEach(() => {
                wrapper = mount(
                    <FieldWithValidation useTooltip>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should render the tooltip, but without content', () => {
                wrapper.update();

                expect(wrapper.find('Input')).toHaveLength(1);
                expect(wrapper.find('Tooltip')).toHaveLength(1);
                expect(wrapper.find('Text')).toHaveLength(0);

                wrapper.find('input').simulate('mouseover');
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(0);
            });
        });
        describe('when error message is defined', () => {
            const message = 'invalid field';
            beforeEach(() => {
                wrapper = mount(
                    <FieldWithValidation errorMessage={message} useTooltip>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should set context to bad on child', () => {
                expect(wrapper.find('Input').prop('context')).toBe('bad');
            });
            it('should render the message on mouse over', () => {
                expect(wrapper.find('Text')).toHaveLength(0);
                expect(wrapper.find('Tooltip')).toHaveLength(1);
                expect(wrapper.find('Tooltip').prop('content')).toBe(message);

                wrapper.find('input').simulate('mouseover');
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(1);
                expect(wrapper.find('div[data-popup="true"]').text()).toBe(message);
            });
            it('should remove the message on mouse leave', () => {
                wrapper.find('input').simulate('mouseover');
                expect(wrapper.find('div[data-popup="true"]').text()).toBe(message);

                wrapper.find('input').simulate('mouseleave');
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(0);
            });
            it('should render the message when field is focused', () => {
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(0);

                wrapper.find('input').simulate('focus');
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(1);
                expect(wrapper.find('div[data-popup="true"]').text()).toBe(message);
            });
            it('should remove the message when field is blurred', () => {
                wrapper.find('input').simulate('focus');
                expect(wrapper.find('div[data-popup="true"]').text()).toBe(message);

                wrapper.find('input').simulate('blur');
                expect(wrapper.find('div[data-popup="true"]')).toHaveLength(0);
            });
        });
    });
});
