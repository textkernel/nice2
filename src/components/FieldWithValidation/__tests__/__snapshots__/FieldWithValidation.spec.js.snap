// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWithValidation when rendering message as text with no error message passed should simply render the child as it is 1`] = `
<FieldWithValidation
  useTooltip={false}
>
  <Input
    disabled={false}
    isBlock={false}
    size="normal"
    type="text"
  >
    <input
      className="Input"
      disabled={false}
      type="text"
    />
  </Input>
</FieldWithValidation>
`;
